xls/modules/rle/rle.x

```diff
@@ -0,0 +1,465 @@
+
+    init {()}
+
+    config (
+        input_r: chan<EncInData<32>> in,
+        output_s: chan<EncOutData<32, 2>> out,
+    ) {
+        spawn RLEEnc<u32:32, u32:2>(input_r, output_s);
+        ()
+    }
+
+    next (tok: token, state: ()) {
+        ()
+    }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH = u32:2;
+
+//const MAX_COUNT = {std::unsigned_max_value<TEST_COUNT_WIDTH>()};
+const MAX_COUNT = u2:3;
+
+type TestSymbol = bits[TEST_SYMBOL_WIDTH];
+type TestCount = bits[TEST_COUNT_WIDTH];
+type TestEncInData = EncInData<TEST_SYMBOL_WIDTH>;
+type TestEncOutData = EncOutData<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>;
+
+// Tightly coupled testing approach
+
+#[test_proc]
+proc RLEEncTightTester {
+    terminator: chan<bool> out;
+    enc_input_s: chan<TestEncInData> out;
+    enc_output_r: chan<TestEncOutData> in;
+
+    init {()}
+    config (terminator: chan<bool> out) {
+        let (enc_input_s, enc_input_r) = chan<TestEncInData>;
+        let (enc_output_s, enc_output_r) = chan<TestEncOutData>;
+
+        spawn RLEEnc<TEST_SYMBOL_WIDTH, TEST_COUNT_WIDTH>(enc_input_r, enc_output_s);
+        (terminator, enc_input_s, enc_output_r)
+    }
+    next (tok: token, state:()) {
+
+        // Clear on `last`
+        let trans_send: TestEncInData[4] = [
+            TestEncInData {symbol: TestSymbol:0xA, last: false},
+            TestEncInData {symbol: TestSymbol:0xA, last: false},
+            TestEncInData {symbol: TestSymbol:0xA, last: false},
+            TestEncInData {symbol: TestSymbol:0xB, last: true},
+        ];
+        let tok = for ((counter, data_in), tok): ((u32, TestEncInData) , token) in enumerate(trans_send) {
+            let tok = send(tok, enc_input_s, data_in);
+            let _ = trace_fmt!("Sent {} transactions, symbol: 0x{:x}, last: {}",
+                counter, data_in.symbol, data_in.last);
+            (tok)
+        }(tok);
+        let trans_recv: TestEncOutData[2] = [
+            TestEncOutData {symbol: TestSymbol:0xA, count: TestCount:0x3, last:false},
+            TestEncOutData {symbol: TestSymbol:0xB, count: TestCount:0x1, last:true},
+        ];
+        let (tok, _, _) = for ((counter, data_out), (tok, can_repeat, prev_symbol)): ((u32, TestEncOutData) , (token, bool, TestSymbol)) in enumerate(trans_recv) {
+            let (tok, enc_output) = recv(tok, enc_output_r);
+            let _ = trace_fmt!(
+                "Received {} transactions, symbol: 0x{:x}, count: {}, last: {}",
+                counter, enc_output.symbol, enc_output.count, enc_output.last
+            );
+            let _ = assert_eq(enc_output, data_out);
+            let max_count = std::unsigned_max_value<TEST_COUNT_WIDTH>();
+
+            let _ = assert_eq(enc_output.count > TestCount:0, true);
+            let _ = assert_eq(enc_output.count <= max_count, true);
+            let data_repeats = prev_symbol == enc_output.symbol;
+            let _ = assert_eq(data_repeats && can_repeat, data_repeats);
```

Comment:
I don't think this can even fail for the given test data: we never got over max_count, and we never iterate past the last symbol.